examples/cfq/models.py

```diff
@@ -111,24 +201,41 @@ def apply(self,
                               ATTENTION_SIZE,
                               name='keys',
                               bias=False)
+    
+    batch_size = encoder_hidden_states.shape[0]
+    hidden_size = encoder_hidden_states.shape[-1]
+    h_dropout_masks = self.create_dropout_masks(num_masks=num_layers,
+                                                shape=(batch_size, hidden_size),
+                                                dropout_rate=horizontal_dropout_rate)
+    v_dropout_masks = self.create_dropout_masks(num_masks=num_layers-1,
+                                                shape=(batch_size, hidden_size),
+                                                dropout_rate=vertical_dropout_rate)
 
     def decode_step_fn(carry, x):
-      rng, (c, h), last_prediction = carry
+      rng, multilayer_lstm_output, last_prediction = carry
+      previous_states, h = multilayer_lstm_output
       carry_rng, categorical_rng = jax.random.split(rng, 2)
-      if not teacher_force:
+      if not train:
         x = last_prediction
       x = shared_embedding(x)
       dec_prev_state = jnp.expand_dims(h, 1)
       attention = mlp_attention(dec_prev_state, projected_keys,
                                 encoder_hidden_states, attention_mask)
-      lstm_state, y = lstm_cell((c, attention), x)
+      previous_states[0] = (previous_states[0][0],attention)
+      
+      states, y = multilayer_lstm_cell(horizontal_dropout_masks=h_dropout_masks,
+                                       vertical_dropout_masks=v_dropout_masks,
+                                       input=x,
+                                       previous_states=previous_states)
       logits = projection(y)
       predicted_tokens = jax.random.categorical(categorical_rng, logits)
       predicted_tokens_uint8 = jnp.asarray(predicted_tokens, dtype=jnp.uint8)
-      return (carry_rng, lstm_state,
+      return (carry_rng, (states, y),
```

Comment:
Why do you pass `(states, y)`, while in the previous version you only passed `state`?

Reply:
I need to pass the state at each layer, as I am keeping track of them outside of the MultilayerLSTM module. Also, I need to pass around y as well because since I have dropout the hidden state and the output now differ